src/qt/trafficgraphwidget.cpp

```diff
@@ -1,257 +1,305 @@
+    painter.setPen(axisCol);
+    painter.drawLine(XMARGIN, YMARGIN + h, width() - XMARGIN, YMARGIN + h);
+
+    // decide what order of magnitude we are
+    int base = floor(log10(fMax));
+    float val = pow(10.0f, base);
+
+    const QString units     = tr("KB/s");
+    const float yMarginText = 2.0;
+
+    // draw lines
+    painter.setPen(axisCol);
+    painter.drawText(XMARGIN, YMARGIN + h - h * val / fMax-yMarginText, QString("%1 %2").arg(val).arg(units));
+    for(float y = val; y < fMax; y += val) {
+        int yy = YMARGIN + h - h * y / fMax;
+        painter.drawLine(XMARGIN, yy, width() - XMARGIN, yy);
+    }
+
+    // if we drew 3 or fewer lines, break them up at the next lower order of magnitude
+    if(fMax / val <= 3.0f) {
+        axisCol = axisCol.darker();
+        val = pow(10.0f, base - 1);
+        painter.setPen(axisCol);
+        painter.drawText(XMARGIN, YMARGIN + h - h * val / fMax-yMarginText, QString("%1 %2").arg(val).arg(units));
+        int count = 1;
+        for(float y = val; y < fMax; y += val, count++) {
+            // don't overwrite lines drawn above
+            if(count % 10 == 0)
+                continue;
+            int yy = YMARGIN + h - h * y / fMax;
+            painter.drawLine(XMARGIN, yy, width() - XMARGIN, yy);
+        }
+    }
+
+    QPainterPath samplesPathIn;
+    paintPath(samplesPathIn, vTimespanSamplesIn, nTimespanHeadIndex);
+    painter.fillPath(samplesPathIn, QColor(0, 255, 0, 128));
+    painter.setPen(Qt::green);
+    painter.drawPath(samplesPathIn);
+
+    QPainterPath samplesPathOut;
+    paintPath(samplesPathOut, vTimespanSamplesOut, nTimespanHeadIndex);
+    painter.fillPath(samplesPathOut, QColor(255, 0, 0, 128));
+    painter.setPen(Qt::red);
+    painter.drawPath(samplesPathOut);
+}
+
+void TrafficGraphWidget::updateTimespanRates()
+{
+    if (!clientModel) {
+        return;
+    }
+
+    int64_t currentTimeMs = GetTimeMillis();
+    int64_t elapsedMs = currentTimeMs - nTimespanLastSampleTimeMs;
+    if (elapsedMs <= 0) {
+        return;
+    }
+
+    quint64 bytesIn = clientModel->getTotalBytesRecv();
+    quint64 bytesOut = clientModel->getTotalBytesSent();
+
+    float inRate = (bytesIn - nTimespanLastBytesIn) / 1000.0f * (1000.0f / elapsedMs);
+    float outRate = (bytesOut - nTimespanLastBytesOut) / 1000.0f * (1000.0f / elapsedMs);
+
+    nTimespanLastBytesIn = bytesIn;
+    nTimespanLastBytesOut = bytesOut;
+    nTimespanLastSampleTimeMs = currentTimeMs;
+
+    vTimespanSamplesIn[nTimespanHeadIndex] = inRate;
+    vTimespanSamplesOut[nTimespanHeadIndex] = outRate;
+
+    nTimespanHeadIndex++;
+    if (nTimespanHeadIndex >= NUM_DISPLAYED_SAMPLES) {
+        nTimespanHeadIndex = 0;
+    }
+
+    fMax = 0.0f;
+    for (int i = 0; i < vTimespanSamplesIn.size(); i++)
+    {
+        fMax = std::max(fMax, vTimespanSamplesIn[i]);
+    }
+    for (int i = 0; i < vTimespanSamplesOut.size(); i++)
+    {
+        fMax = std::max(fMax, vTimespanSamplesOut[i]);

```

Comment:
I keep looking for a way to reduce the complexity here, and the best I can do is combine the traversal. If both of these vectors are the same size, can we get `O(n)` instead of `O(2n)`?

If it's clearer, loop from `0` to `NUM_DISPLAYED_SAMPLES` perhaps?

Reply:
I couldn't think of a situation where the size differs, but to be defensive, could take the `std::min()`? eg:

```cpp
    const int nInSampleSize = std::min(vTimespanSamplesIn.size(), vTimespanSamplesOut.size());
    for (int i = 0; i < nInSampleSize; i++)
    {
        fMax = std::max(std::max(fMax, vTimespanSamplesIn[i]), vTimespanSamplesOut[i]);
    }

```

Reply:
Ok to limit the minimum (should never happen, and would throw an error, but, if it's fed bad sizes, sure, lets trim it) and keep code analysis tools happy.   Loop combining though definitely all for. So that is done.